proto/v1beta1/intoto.proto

```diff
@@ -0,0 +1,135 @@
+// Copyright 2020 The Grafeas Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package grafeas.v1beta1.intoto;
+
+option go_package = "github.com/grafeas/grafeas/proto/v1beta1/intoto_go_proto";
+option java_multiple_files = true;
+option java_package = "io.grafeas.v1beta1.intoto";
+option objc_class_prefix = "GRA";
+
+import "google/protobuf/struct.proto";
+
+// This file defines the proto that matches the schema of in-toto metadata.
+// in-toto is a framework to protect a software supply chain. You can read more
+// about in-toto at https://in-toto.io/ and you can find the current version of
+// the in-toto specification at
+// https://github.com/in-toto/docs/blob/master/in-toto-spec.md
+// There are two types of in-toto metadata - layouts and links. Layouts define
+// the structure of the software supply chain, while links are attestations of
+// performing a step in the software supply chain.
+
+// This contains the fields corresponding to the definition of a software supply
+// chain step in an in-toto layout. This information goes into a Grafeas note.
+message InToto {
+  // This field identifies the name of the step in the supply chain.
+  string step_name = 1;
+
+  // This field contains the public keys that can be used to verify the
+  // signatures on the step metadata.
+  repeated SigningKey signing_keys = 2;
+
+  // The following fields contain in-toto artifact rules identifying the
+  // artifacts that enter this supply chain step, and exit the supply chain
+  // step, i.e. materials and products of the step.
+  repeated google.protobuf.ListValue expected_materials = 3;
+  repeated google.protobuf.ListValue expected_products = 4;
```

Comment:
The previous thread on this is locked for some reason, so continuing the conversation here, [context](https://github.com/grafeas/grafeas/pull/391#discussion_r369306952):

So to make sure we're talking about the same thing, here's my suggestion:

```
message Material {
  repeated string value = 1; // there's probably a better nam than `value` for this, so I defer to you
}

message InToto {
  ...
 repeated Material expected_materials = 3;
 repeated Material expected_products = 4;
}
}
```

You could also next the `Material` definition [inside the InToto proto](https://developers.google.com/protocol-buffers/docs/proto3#nested), right before the field.

I'm guessing the main concern here is how it's generated in JSON, correct? How would it be represented with the proposal above vs what you need it to look like?

Reply:
Yes. I built this locally and the metadata needs to look like:
```
        "expectedProducts": [
          {
            "value": [
              "CREATE",
              "demo-project/foo.py"
            ]
          },
          {
            "value": [
              "DISALLOW",
              "*"
            ]
          }
        ],
```
instead of:
```
        "expectedProducts": [
          [
            "CREATE",
            "demo-project/foo.py"
          ],
          [
            "DISALLOW",
            "*"
          ]
        ],
```

This is what led me to use `ListValue` to allow for nested lists _without_ an intermediate map, so to speak.

Reply:
Got it! In this case, PTAL at my suggestion from a few minutes ago about structuring the `Material` proto message with a map as the field.

Reply:
Resolving this to condense the various threads about the same thing.